pkg/controller/redissnapshot/redissnapshot_controller.go

```diff
@@ -0,0 +1,246 @@
+			return reconcile.Result{}, updateErr
+		}
+		return reconcile.Result{}, err
+	}
+	if stratCfg.Region == "" {
+		stratCfg.Region = croAws.DefaultRegion
+	}
+
+	// create the credentials to be used by the aws resource providers, not to be used by end-user
+	providerCreds, err := r.CredentialManager.ReconcileProviderCredentials(ctx, redisCr.Namespace)
+	if err != nil {
+		errMsg := "failed to reconcile elasticache credentials"
+		if updateErr := resources.UpdateSnapshotPhase(ctx, r.client, instance, croType.PhaseFailed, croType.StatusMessage(errMsg)); updateErr != nil {
+			return reconcile.Result{}, updateErr
+		}
+		return reconcile.Result{}, errorUtil.Wrap(err, errMsg)
+	}
+
+	// setup aws elasticache cluster sdk session
+	cacheSvc := elasticache.New(session.Must(session.NewSession(&aws.Config{
+		Region:      aws.String(stratCfg.Region),
+		Credentials: credentials.NewStaticCredentials(providerCreds.AccessKeyID, providerCreds.SecretAccessKey, ""),
+	})))
+
+	// create snapshot of primary node
+	phase, msg, err := r.createSnapshot(ctx, cacheSvc, instance, redisCr)
+	if updateErr := resources.UpdateSnapshotPhase(ctx, r.client, instance, phase, msg); updateErr != nil {
+		return reconcile.Result{}, updateErr
+	}
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 60}, nil
+}
+
+func (r *ReconcileRedisSnapshot) createSnapshot(ctx context.Context, cacheSvc elasticacheiface.ElastiCacheAPI, snapshot *integreatlyv1alpha1.RedisSnapshot, redis *integreatlyv1alpha1.Redis) (croType.StatusPhase, croType.StatusMessage, error) {
+	// generate snapshot name
+	snapshotName, err := croAws.BuildTimestampedInfraNameFromObjectCreation(ctx, r.client, snapshot.ObjectMeta, croAws.DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to generate snapshot name"
+		return croType.PhaseFailed, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// generate cluster name
+	clusterName, err := croAws.BuildInfraNameFromObject(ctx, r.client, redis.ObjectMeta, croAws.DefaultAwsIdentifierLength)
+	if err != nil {
+		errMsg := "failed to get cluster name"
+		return croType.PhaseFailed, croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	// check snapshot exists
+	listOutput, err := cacheSvc.DescribeSnapshots(&elasticache.DescribeSnapshotsInput{
+		SnapshotName: aws.String(snapshotName),
+	})
+	var foundSnapshot *elasticache.Snapshot
+	for _, c := range listOutput.Snapshots {
+		if *c.SnapshotName == snapshotName {
+			foundSnapshot = c
+			break
+		}
+	}
+
+	// get replication group
+	cacheOutput, err := cacheSvc.DescribeReplicationGroups(&elasticache.DescribeReplicationGroupsInput{
+		ReplicationGroupId: aws.String(clusterName),
+	})
+
+	// ensure replication group is available
+	if *cacheOutput.ReplicationGroups[0].Status != "available" {
```

Comment:
Is it possible to have no elements in the ReplicationGroups list? I'd assume no, but just want to clarify if we need a check here